plugins/transformations/seq2array/srcCxx/Seq2ArrayTransformation.cxx

```diff
@@ -0,0 +1,481 @@
+                    + ">, output name <" + output_union_type.name() + ">: "
+                    + std::to_string(input_union_type.member_count())
+                    + " != " + std::to_string(output_union_type.member_count()));
+            throw std::runtime_error(error);
+        }
+
+        if (input_union_type.discriminator() != output_union_type.discriminator()){
+            std::string error("Error: different discriminator of input and output "
+                    "unions. Input union <" + input_union_type.name()
+                    + "> discriminator name <" + input_union_type.discriminator().name()
+                    + "> is different of output union <" + output_union_type.name()
+                    + "> discriminator name <" + output_union_type.discriminator().name()
+                    + ">.");
+            throw std::runtime_error(error);
+        }
+
+        if (input_union_type == output_union_type) {
+            return true;
+        } else {
+            // inspect element by element since it may be compatible
+            // (output with arrays and input with sequences)
+            for (auto member : input_union_type.members()) {
+                switch (member.type().kind().underlying()) {
+                case TypeKind::STRUCTURE_TYPE:
+                case TypeKind::SEQUENCE_TYPE:
+                case TypeKind::ARRAY_TYPE:
+                case TypeKind::UNION_TYPE:
+                    is_compatible = is_compatible && are_type_compatible(
+                            member.type(),
+                            output_union_type.member(member.name()).type());
+                    if (!is_compatible) {
+                        std::string error("Error: input member <"
+                                + member.name().to_std_string()
+                                + ">, not compatible with output name <"
+                                + output_union_type.member(member.name()).name().to_std_string()
+                                + ">");
+                        throw std::runtime_error(error);
+                    }
+                break;
+                // primitive type
+                default:
+                    is_compatible = is_compatible
+                            && (member.type() == output_union_type.member(member.name()).type());
+                    if (!is_compatible) {
+                        std::string error("Error: input member <"
+                                + member.name().to_std_string()
+                                + ">, not compatible with output name <"
+                                + output_union_type.member(member.name()).name().to_std_string()
+                                + ">");
+                        throw std::runtime_error(error);
+                    }
+                    break;
+                }
+
+                if (!is_compatible) {
+                    break;
+                }
+            }
```

Comment:
This block is pretty much identical to the handling of struct types, so it would be great to avoid the duplication and refactor the code so that the logic is shared between the two (either by refactoring the check into a separate function, or by letting this function "fall through" onto this logic only in the case of struct and union types).

Reply:
done